tornettools/generate.py

```diff
@@ -133,14 +141,57 @@ def __get_scaled_tgen_server_bandwidth_kbit(args):
     scaled_bw = scaled_client_bw * n_clients_per_server
     return scaled_bw
 
-def __server(args, server):
+def __filter_nodes(network, ip_address_hint, country_code_hint):
+    all_nodes = [{'id': node[0], **node[1]} for node in network.nodes(data=True)]
+
+    if ip_address_hint is not None and not ip_address_hint.is_global:
+        # ignore the hint if the IP address is not global
+        logging.debug(f"Ignoring non-global address {ip_address_hint}")
+        ip_address_hint = None
+
+    if country_code_hint is not None:
+        # normalize the country code
+        country_code_hint = country_code_hint.casefold()
+
+    # are there any nodes with the same ip address?
+    ip_match_found = any('ip_address' in node and IPv4Address(node['ip_address']) == ip_address_hint for node in all_nodes)
+
+    if ip_match_found:
+        # get all nodes with exact IP matches, regardless of the country code
+        candidate_nodes = [node for node in all_nodes if 'ip_address' in node and IPv4Address(node['ip_address']) == ip_address_hint]
+    else:
+        # get all nodes with the same country code
+        candidate_nodes = [node for node in all_nodes if 'country_code' in node and node['country_code'].casefold() == country_code_hint]
+
+        # if no node had the same country code, use all nodes
+        if len(candidate_nodes) == 0:
+            candidate_nodes = [node for node in all_nodes]
+
+        any_ip_found = any('ip_address' in node for node in candidate_nodes)
+
+        # if a node has an IP address and we were given an IP hint, perform longest prefix matching
+        if any_ip_found and ip_address_hint is not None:
+            # exclude nodes without an IP address
+            candidate_nodes = [node for node in candidate_nodes if 'ip_address' in node]
+
+            longest_prefix_match = lambda ip_1, ip_2: ~(int(ip_1)^int(ip_2))
+
+            # get the single node with the longest prefix match
+            candidate_nodes = [max(candidate_nodes, key=lambda node: longest_prefix_match(IPv4Address(node['ip_address']), ip_address_hint))]
```

Comment:
What if two nodes have the same prefix length to the target ip address hint? For example, what if the hint address is 2 and both nodes have the same longest prefix matched ip address of 3?

Reply:
Do you mean if two nodes have the same IP address? I think the only case where `longest_prefix_match(A, X) == longest_prefix_match(A, Y)` is if `X == Y` (using the formula `longest_prefix_match(I, J) = ~(I^J)` that shadow uses).

Reply:
Right. I suppose it's possible that `X == Y`, though this probably doesn't happen in the atlas graph so maybe not worth dealing with? I think your current code would just choose either `X` or `Y` but I'm not clear on how that choice would be made.

Reply:
The `max()` function will return the first occurrence, which matches how shadow works:

```c
in_addr_t match = ~(vertexIP ^ ip);
if(match > bestMatch || bestMatch == 0) {
    bestMatch = match;
    bestIP = vertexIP;
    bestVertexIndexPtr = vertexIndexPtr;
}
```

But this is easy to change, so I'll make it return all matching candidates.

Reply:
Added.